packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -630,15 +662,23 @@ class TextPainter {
   /// To set the text style, specify a [TextStyle] when creating the [TextSpan]
   /// that you pass to the [TextPainter] constructor or to the [text] property.
   void paint(Canvas canvas, Offset offset) {
-    assert(() {
-      if (_needsLayout) {
-        throw FlutterError(
-          'TextPainter.paint called when text geometry was not yet calculated.\n'
-          'Please call layout() before paint() to position the text before painting it.',
-        );
-      }
-      return true;
-    }());
+    final double? minWidth = _lastMinWidth;
+    final double? maxWidth = _lastMaxWidth;
+    if (_paragraph == null || minWidth == null || maxWidth == null) {
+      throw StateError(
+        'TextPainter.paint called when text geometry was not yet calculated.\n'
+        'Please call layout() before paint() to position the text before painting it.',
+      );
+    }
+
+    if (_needsPaint) {
```

Comment:
the name _needsPaint seems confusing to me. The TextPainter is drawing/painting the paragraph in line 682 even when this is false.

Reply:
_needsRebuildDuringPaint, maybe?

Reply:
`_rebuildParagraphForLayout` (a getter) and `_rebuildParagraphForPaint` ?